crates/adapters/src/controller/stats.rs

```diff
@@ -955,6 +1165,203 @@ impl StepResults {
     }
 }
 
+const MAX_TOKENLIST_LEN: usize = 1_000_000;
+
+/// Tracks the set of tokens issued by the connector and the state needed to report
+/// completion status for these tokens.
+///
+/// A token is the number of records ingested by the connector at the time the token
+/// was requested.  When all records up to this offset are pushed from the connector's
+/// queue into the circuit, we record the corresponding index (global offset) in the
+/// globally ordered stream of input to the circuit.  We can then determine whether
+/// all records associated with the token have been fully processed by comparing this
+/// offset with the `total_completed_records` counter.
+///
+/// # Garbage collection
+///
+/// A new token is generated on each /completion_token API call, as well as for each
+/// HTTP /ingress request. We therefore need to be careful to bound the number of tokens
+/// in the list.
+///
+/// We do this by removing all except one token whose global offset is `<= total_completed_records`
+/// from the list.  This bounds the number of tokens to the size of the connectors
+/// input queue + the number of inputs currently in the circuit of buffered in the output
+/// connectors. Which can be quite large. Therefore we additionally bound the list to `MAX_TOKENLIST_LEN`
+/// tokens, dropping the oldest token when queue is full. This is safe (i.e., we can never
+/// report an operation that hasn't been completed as completed), but it can delay the
+/// completion notification. In the worst case, if there are always `> MAX_TOKENLIST_LEN` incomplete
+/// tokens in the queue, we will never issue any completion notifications, but that's
+/// probably ok, as this means that no one's waiting for these tokens.
+#[derive(Serialize)]
+#[repr(transparent)]
+struct TokenList(Mutex<TokenListInner>);
+
+#[derive(Serialize)]
+struct TokenListInner {
+    /// List of (token, global offset) mappings.
+    token_list: VecDeque<(u64, Option<u64>)>,
+}
+
+impl TokenListInner {
+    fn new() -> Self {
```

Comment:
this could be abstracted into a utility class "BoundedMonotoneResource", in case you find other uses for it.
It prevents some DDOS attacks